From f02806315c9d468931489e0d7d9ee9cac738fcf5 Mon Sep 17 00:00:00 2001
From: Sergio Lopez <slp@redhat.com>
Date: Thu, 19 May 2022 22:42:01 +0200
Subject: [PATCH 10/21] tsi: allow hijacking sockets (tsi_hijack)

Add a kernel command line option (tsi_hijack) enabling users to
request the kernel to hijack AF_INET(SOCK_STREAM || SOCK_DGRAM)
sockets to turn them into TSI sockets, and another one
(tsi_hijack_unix) to do the same thing for AF_UNIX sockets.

Signed-off-by: Sergio Lopez <slp@redhat.com>
---
 net/socket.c     | 28 ++++++++++++++++++++++++++++
 net/tsi/af_tsi.c |  2 +-
 2 files changed, 29 insertions(+), 1 deletion(-)

diff --git a/net/socket.c b/net/socket.c
index 4e1ec38c2..30b30bde2 100644
--- a/net/socket.c
+++ b/net/socket.c
@@ -115,6 +115,11 @@ unsigned int sysctl_net_busy_read __read_mostly;
 unsigned int sysctl_net_busy_poll __read_mostly;
 #endif
 
+#ifdef CONFIG_TSI
+bool tsi_hijack = false;
+bool tsi_hijack_unix = false;
+#endif
+
 static ssize_t sock_read_iter(struct kiocb *iocb, struct iov_iter *to);
 static ssize_t sock_write_iter(struct kiocb *iocb, struct iov_iter *from);
 static int sock_mmap(struct file *file, struct vm_area_struct *vma);
@@ -1490,6 +1495,11 @@ int sock_wake_async(struct socket_wq *wq, int how, int band)
 }
 EXPORT_SYMBOL(sock_wake_async);
 
+#ifdef CONFIG_TSI
+core_param(tsi_hijack, tsi_hijack, bool, 0644);
+core_param(tsi_hijack_unix, tsi_hijack_unix, bool, 0644);
+#endif
+
 /**
  *	__sock_create - creates a socket
  *	@net: net namespace
@@ -1560,6 +1570,24 @@ int __sock_create(struct net *net, int family, int type, int protocol,
 		request_module("net-pf-%d", family);
 #endif
 
+#ifdef CONFIG_TSI
+	if (!kern && (type == SOCK_STREAM || type == SOCK_DGRAM)) {
+		if (family == AF_INET && tsi_hijack) {
+			pr_debug("%s - tsi: hijacking AF_INET socket\n",
+				current->comm);
+			family = AF_TSI;
+		} else if (family == AF_INET6 && tsi_hijack) {
+			pr_debug("%s - tsi: hijacking AF_INET6 socket\n",
+				current->comm);
+			family = AF_TSI6;
+		} else if (family == AF_UNIX && tsi_hijack_unix) {
+			pr_debug("%s - tsi: hijacking AF_UNIX socket\n",
+				current->comm);
+			family = AF_TSIU;
+		}
+	}
+#endif
+
 	rcu_read_lock();
 	pf = rcu_dereference(net_families[family]);
 	err = -EAFNOSUPPORT;
diff --git a/net/tsi/af_tsi.c b/net/tsi/af_tsi.c
index e266f9684..a8940e3e9 100644
--- a/net/tsi/af_tsi.c
+++ b/net/tsi/af_tsi.c
@@ -547,7 +547,7 @@ static int tsi_accept(struct socket *sock, struct socket *newsock,
 	struct socket *csocket;
 	struct tsi_sock *tsk;
 	struct tsi_sock *newtsk;
-	struct socket *nsock;
+	struct socket *nsock = NULL;
 	struct sock *sk;
 	int err;
 
-- 
2.51.1

